File: ex10_c_dev_kit/examples/access_read_write_example.c
/*****************************************************************************
 *                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      *
 *                                                                           *
 * This source code is the property of Impinj, Inc. Your use of this source  *
 * code in whole or in part is subject to your applicable license terms      *
 * from Impinj.                                                              *
 * Contact support@impinj.com for a copy of the applicable Impinj license    *
 * terms.                                                                    *
 *                                                                           *
 * (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               *
 *                                                                           *
 *****************************************************************************/

#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "board/time_helpers.h"
#include "ex10_api/application_registers.h"
#include "ex10_api/board_init.h"
#include "ex10_api/event_fifo_printer.h"
#include "ex10_api/ex10_helpers.h"
#include "ex10_api/ex10_reader.h"
#include "ex10_api/gen2_tx_command_manager.h"
#include "ex10_api/rf_mode_definitions.h"
#include "ex10_api/trace.h"
#include "ex10_api/application_register_definitions.h"

#ifdef randy
//**************************************************************
// Add SDD logging code here as a test. 
// What determines the file name
// enable power control loop logs​
void enable_sdd(const struct Ex10Interfaces ex10_iface)
{
    struct LogEnablesFields log_enables = {0};
    log_enables.op_logs                 = true;
    log_enables.ramping_logs            = true;
    log_enables.lmac_logs               = true;
    log_enables.insert_fifo_event_logs  = true;
    log_enables.host_irq_logs           = true;
    log_enables.regulatory_logs         = true;
    const uint8_t log_speed_mhz         = 12;
    ex10_iface.reader->enable_sdd_logs(log_enables, log_speed_mhz);
}

//disable undesired fields​
//    memset(log_enables_reg_value, 0, sizeof (log_enables_reg_value));​
    
//enable desired field(s)​
    log_enables.ramping_logs = true;​

//set clock frequency​
    uint8_t log_sdd_clk_MHz = 4;​

​//configure fields and clock frequency​
    ex10_iface.ops->enable_sdd_logs(log_enables_reg_value, log_sdd_clk_MHz);

//**************************************************************
#endif 

/* Settings used when running this example */
static uint32_t const inventory_duration_ms = 500;  // Duration in milliseconds
static uint8_t const  antenna               = 1;
static uint16_t const rf_mode               = mode_148;
static uint16_t const transmit_power_cdbm   = 3000;
static uint8_t const  initial_q             = 2;
static bool const     dual_target           = true;
static uint8_t const  session               = 0;
static bool const     tag_focus_enable      = false;
static bool const     fast_id_enable        = false;


/* Global state */
struct Ex10Interfaces  ex10_iface                     = {0u};
struct InfoFromPackets packet_info                    = {0u, 0u, 0u, 0u, {0u}};
struct Gen2CommandSpec access_cmds[MaxTxCommandCount] = {0u};
bool                   halted_enables[MaxTxCommandCount] = {0u};

/* Gen2 parameters for this example */
struct WriteCommandArgs write_args = {
    .memory_bank  = User,
    .word_pointer = 0u,
    .data         = 1234u,  // Modify value before use
};

struct ReadCommandArgs read_args = {
    .memory_bank  = User,
    .word_pointer = 0u,
    .word_count   = 1u,
};

/**
 * Before starting inventory, setup gen2 sequence to write a random 16-bit
 * value to user memory bank offset 0, and then read back the word at that
 * location
 */
static void setup_gen2_write_read_sequence(uint16_t data_word)
{
    /* Setup read and write commands ahead of time */

    struct Ex10Gen2TxCommandManager const* g2tcm =
        get_ex10_gen2_tx_command_manager();
    g2tcm->clear_local_sequence();

    write_args.data                  = data_word;
    struct Gen2CommandSpec write_cmd = {
        .command = Gen2Write,
        .args    = &write_args,
    };

    struct Gen2TxCommandManagerError error =
        g2tcm->encode_and_append_command(&write_cmd, 0);
    assert(false == error.error_occurred);
    assert(error.current_index == 0u);
    halted_enables[error.current_index] = true;
    access_cmds[error.current_index]    = write_cmd;

    struct Gen2CommandSpec read_cmd = {
        .command = Gen2Read,
        .args    = &read_args,
    };

    error = g2tcm->encode_and_append_command(&read_cmd, 1);
    assert(false == error.error_occurred);
    assert(error.current_index == 1u);
    halted_enables[error.current_index] = true;
    access_cmds[error.current_index]    = read_cmd;

    g2tcm->write_sequence();

    g2tcm->write_halted_enables(halted_enables, MaxTxCommandCount);
}

static int run_inventory_rounds(enum SelectType const select_type)
{
    struct InventoryRoundControlFields inventory_config = {
        .initial_q            = initial_q,
        .max_q                = initial_q,
        .min_q                = initial_q,
        .num_min_q_cycles     = 1,
        .fixed_q_mode         = true,
        .q_increase_use_query = false,
        .q_decrease_use_query = false,
        .session              = session,
        .select               = select_type,
        .target               = 0,
        .halt_on_all_tags     = true,
        .tag_focus_enable     = tag_focus_enable,
        .fast_id_enable       = fast_id_enable,
    };

    struct InventoryRoundControl_2Fields const inventory_config_2 = {
        .max_queries_since_valid_epc = 0};

    struct InventoryHelperParams ihp = {
        .antenna               = antenna,
        .rf_mode               = rf_mode,
        .tx_power_cdbm         = transmit_power_cdbm,
        .inventory_config      = &inventory_config,
        .inventory_config_2    = &inventory_config_2,
        .send_selects          = false,
        .remain_on             = false,
        .dual_target           = dual_target,
        .inventory_duration_ms = inventory_duration_ms,
        .packet_info           = &packet_info,
        .verbose               = true};

    if (ex10_iface.helpers->simple_inventory(&ihp))
    {
        return -1;
    };

    if (!packet_info.total_singulations)
    {
        printf("No tags found in inventory\n");
        return -1;
    }

    return 0;
}

/**
 * Return pointer to next enabled Gen2 Access Command
 * NOTE: Assumes at least one valid access command at index 0
 */
static struct Gen2CommandSpec const* next_cmd(void)
{
    static size_t                 cmd_index = 0u;
    struct Gen2CommandSpec const* next      = &access_cmds[0u];

    for (size_t iter = cmd_index; iter < 10u; iter++)
    {
        if (halted_enables[iter])
        {
            next      = &access_cmds[iter];
            cmd_index = iter + 1u;
            break;
        }
    }
    return next;
}

/**
 * Run inventory round, halt on first tag, execute gen2 sequence
 */
static int run_access_read_write_seq(void)
{
    write_args.data = rand() % UINT16_MAX;
    printf("Number to write %X\n",write_args.data);
    setup_gen2_write_read_sequence(write_args.data);

    if (run_inventory_rounds(SelectAll) == -1)
    {
        ex10_iface.reader->stop_transmitting();
        return -1;
    }

    /* Should be halted on a tag now */
    bool halted = ex10_iface.helpers->inventory_halted();
    assert(halted && "Failed to halt on a tag");

    printf("Sending write and read commands\n");

    /* Trigger stored Gen2 sequence */
    ex10_iface.ops->send_gen2_halted_sequence();

    /* Wait for Gen2Transaction packets to be returned */
    uint32_t const   timeout          = 1000;
    uint16_t         reply_words[10u] = {0};
    struct Gen2Reply reply = {.error_code = NoError, .data = reply_words};
    
    size_t gen2_packet_count_expected = 2u;

    uint32_t const start_time = get_ex10_time_helpers()->time_now();
    while (get_ex10_time_helpers()->time_elapsed(start_time) < timeout &&
           gen2_packet_count_expected)
    {
        struct EventFifoPacket const* packet = ex10_iface.reader->packet_peek();
        while (packet != NULL)
        {
            get_ex10_event_fifo_printer()->print_packets(packet);
            if (packet->packet_type == Gen2Transaction)
            {
                gen2_packet_count_expected--;
                reply.error_code = NoError;
                memset(reply_words, 0x00, sizeof(reply_words));

                ex10_iface.gen2_commands->decode_reply(
                    next_cmd()->command, packet, &reply);
                assert(ex10_iface.helpers->check_gen2_error(&reply) == false);
            }
            ex10_iface.reader->packet_remove();
            packet = ex10_iface.reader->packet_peek();
        }
    }

    assert((gen2_packet_count_expected == 0u) &&
           "Did not receive expected number of Gen2Transaction Packets");

    /* Last gen2 reply has result of read command */
    if (reply_words[0] == write_args.data)
    {
        printf("Response 0x%04X from Read command matched what was written\n",
               reply_words[0]);
    }
    else
    {
        printf("Expected: 0x%04x, read: 0x%04x\n", write_args.data, reply_words[0]);

        printf("Gen2Reply for Read:\n");
        printf("  Reply command: %d\n", reply.reply);
        printf("  TagErrorCode: %d\n", reply.error_code);
        printf("  data[0]: 0x%04x\n", reply.data[0]);
    }
    assert(reply_words[0] == write_args.data);

    /* Demonstrate continuing to next tag, not used here. */
    ex10_iface.reader->continue_from_halted(false);

    printf("Ending write+read sequence example\n");

    return 0;
}

int main(void)
{
    tracepoint(pi_ex10sdk, EXEC_start, __FILE__);
    ex10_iface = ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, "FCC");

    int result = ex10_iface.helpers->check_board_init_status(Application);
    if (result == 0)
    {
        result = run_access_read_write_seq();
        if (result == -1)
        {
            printf("Read/write access issue.\n");
            ex10_iface.reader->stop_transmitting();
        }
    }

    ex10_typical_board_teardown();
    tracepoint(pi_ex10sdk, EXEC_end, __FILE__);

    return result;
}

File: ex10_dev_kit/examples/access_read_write_example.py
#############################################################################
#                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      #
#                                                                           #
# This source code is the property of Impinj, Inc. Your use of this source  #
# code in whole or in part is subject to your applicable license terms      #
# from Impinj.                                                              #
# Contact support@impinj.com for a copy of the applicable Impinj license    #
# terms.                                                                    #
#                                                                           #
# (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               #
#                                                                           #
#############################################################################
"""
Halt on the first tag seen, write to it, then read back the value.
"""

from __future__ import (division, absolute_import, print_function,
                        unicode_literals)
from datetime import datetime


from py2c_interface.py2c_python_wrapper import *


# The configuration of the inventory example is set using the definitions below
INVENTORY_DURATION_S = 5            # Duration of inventory operation (seconds)
TRANSMIT_POWER_DBM = 3000           # 30 dBm
RF_MODE = RfModes.mode_5
R807_ANTENNA_PORT = 1               # Which R807 antenna port will be used
INITIAL_Q = 2                       # Q field in the Query command
SELECT_ALL = 0                      # SELECT field in the Query command
SESSION = 0                         # Session field in the Query command
TARGET = 0                          # Target field in the Query command
HALT_ON_ALL_TAGS = True             # If True, the Ex10 will halt on each
                                    # inventoried tag
TAG_FOCUS_ENABLE = False            # Tells a tag to be silent after inventoried
                                    # once.
FAST_ID_ENABLE = False              # Tells a tag to backscatter the TID during
                                    # inventory

# Words to be written and read back from tag
TEST_WORD_0 = 0x55AA

pc = c_uint16()
epc = (c_ubyte * 40)()
crc = c_uint16()
tid = (c_ubyte * 40)()
packet_info = InfoFromPackets(0, 0, 0,
                              TagReadData(pointer(pc),
                                          ctypes.cast(epc, POINTER(c_ubyte)),
                                          0,
                                          pointer(crc),
                                          ctypes.cast(tid, POINTER(c_ubyte)),
                                          0))

cmds = []

def run_access_rw_example():
    # pylint: disable=missing-docstring
    print('Starting read/write example')
    # Init the python to C layer
    py2c = Ex10Py2CWrapper()
    try:
        run_access_read_write(py2c, INITIAL_Q)
    finally:
        print('Ending read/write example')
        py2c.ex10_typical_board_teardown()


def enable_commands(tx_command_manager, enable_index):
    """
    Enable one select at a time for this scenario via the index.
    :param enable_index: The index of the select to enable
    """
    arraytype = ctypes.c_bool * 10
    enable_me = arraytype(0,0,0,0,0,0,0,0,0,0)

    for ind in range(len(enable_index)):
        enable_me[ind] = enable_index[ind]
    enable_p = pointer(enable_me)

    tx_command_manager.write_halted_enables(ctypes.cast(enable_p, c_void_p), 10);


def load_access_commands(tx_command_manager):
    """ Load Gen2 commands into Ex10 buffer so they can be sent later. """
    print('Loading commands into Gen2 command buffer')
    write_args = WriteCommandArgs()
    write_args.memory_bank = MemoryBank.User
    write_args.word_pointer = 0
    write_args.data = TEST_WORD_0

    read_args = ReadCommandArgs()
    read_args.memory_bank = MemoryBank.User
    read_args.word_pointer = 0
    read_args.word_count = 1

    cmds.append(Gen2CommandSpec(Gen2Command.Gen2Write,
                ctypes.cast(pointer(write_args), c_void_p)))
    cmds.append(Gen2CommandSpec(Gen2Command.Gen2Read,
                ctypes.cast(pointer(read_args), c_void_p)))

    # Clear any commands before adding new ones
    tx_command_manager.clear_local_sequence();
    enables = [0]*10
    # Add the command
    curr_error = tx_command_manager.encode_and_append_command(cmds[0], 0);
    assert(False == curr_error.error_occurred);
    enables[curr_error.current_index] = True;
    curr_error = tx_command_manager.encode_and_append_command(cmds[1], 1);
    assert(False == curr_error.error_occurred);
    enables[curr_error.current_index] = True;
    curr_error = tx_command_manager.write_sequence();
    assert(False == curr_error.error_occurred);

    # Enable the command we added
    enable_commands(tx_command_manager, enables)


def run_until_halted(ex10_reader, helper, inventory_config):
    """ Run until halted or timeout trying to find a tag. """
    inventory_config_2 = InventoryRoundControl_2Fields()
    inventory_config_2.max_queries_since_valid_epc = 0

    inv_done = True
    start_time = datetime.now()
    while not packet_info.total_singulations:
        time_delta = (datetime.now() - start_time).total_seconds()
        if time_delta > 5:
            break
        if inv_done:
            inv_done = False
            op_error = ex10_reader.inventory(R807_ANTENNA_PORT,
                                  RF_MODE,
                                  TRANSMIT_POWER_DBM,
                                  pointer(inventory_config),
                                  pointer(inventory_config_2),
                                  None,
                                  0,
                                  True)
            assert op_error.error_occurred == False
        while ex10_reader.packets_available():
            packet = ex10_reader.packet_peek().contents
            if packet.packet_type == EventPacketType.InventoryRoundSummary:
                inv_done = True
            elif packet.packet_type == EventPacketType.TagRead:
                helper.examine_packets(packet, pointer(packet_info))
            ex10_reader.packet_remove()

    # Should be halted on a tag now
    assert packet_info.total_singulations == 1  # Should only be one when halted
    tag_epc = ''.join([str(format(packet_info.access_tag.epc[i], '02x')) for i in range(packet_info.access_tag.epc_length)])
    print('Halted on tag 0x{}'.format(tag_epc))


def wait_for_transaction_report(py2c, ex10_ifaces, expected_reports=1, timeout_s=3):
    ex10_reader = ex10_ifaces.reader
    helper = ex10_ifaces.helpers

    start_time = datetime.now()

    transactions = []
    while len(transactions) < expected_reports:
        while ex10_reader.packets_available():
            packet = ex10_reader.packet_peek().contents
            helper.examine_packets(packet, packet_info)
            helper.print_packets(packet)
            
            if packet.packet_type == EventPacketType.TxRampDown:
                raise RuntimeError(
                    'Modem ramped down while waiting for Gen2 transaction')
            elif packet.packet_type == EventPacketType.Gen2Transaction:
                reply_array = c_uint16 * 10
                reply_words = reply_array()
                reply = Gen2Reply(0, 0, reply_words)
                ex10_ifaces.gen2_commands.decode_reply(cmds[len(transactions)].command, packet, pointer(reply))
                transactions.append(reply)

            ex10_reader.packet_remove()

        if (datetime.now() - start_time).total_seconds() > timeout_s:
            raise RuntimeError('Timed out waiting for Gen2Transaction')
    return transactions


def run_access_read_write(py2c, initial_q):
    # pylint: disable=too-many-locals,too-many-statements
    """ Run inventory for the specified amount of time """
    ex10_ifaces = py2c.ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, 'FCC'.encode('ascii'))
    ex10_reader = ex10_ifaces.reader
    ex10_ops = ex10_ifaces.ops
    helper = ex10_ifaces.helpers
    tx_command_manager = py2c.get_ex10_gen2_tx_command_manager()

    # @todo JIRA PI-22454 [YK FW] EX10 should assert Halted interrupt
    # again when Access commands are done
    ex10_ifaces.protocol.set_event_fifo_threshold(0)

    # Put together configurations for the inventory round
    inventory_config = InventoryRoundControlFields()
    inventory_config.initial_q = initial_q
    inventory_config.max_q = initial_q
    inventory_config.min_q = initial_q
    inventory_config.num_min_q_cycles = 1
    inventory_config.fixed_q_mode = True
    inventory_config.q_increase_use_query = False
    inventory_config.q_decrease_use_query = False
    inventory_config.session = SESSION
    inventory_config.select = SELECT_ALL
    inventory_config.target = TARGET
    inventory_config.halt_on_all_tags = HALT_ON_ALL_TAGS
    inventory_config.fast_id_enable = FAST_ID_ENABLE
    inventory_config.tag_focus_enable = TAG_FOCUS_ENABLE

    print('########## Delayed Write - User Data Test ##########')

    load_access_commands(tx_command_manager)
    
    ex10_ops.send_gen2_halted_sequence();

    run_until_halted(ex10_reader, helper, inventory_config)

    transactions = wait_for_transaction_report(py2c=py2c,
                                               ex10_ifaces=ex10_ifaces,
                                               expected_reports=2)
    assert len(transactions) == 2

    # Transaction reports are received in the same order commands are sent out.
    # Verify the 2 write commands worked.
    write_reply = transactions[0]
    helper.check_gen2_error(write_reply)

    read_reply = transactions[1]
    helper.check_gen2_error(read_reply)
    print('Read Back: 0x{:04x}'.format(read_reply.data[0]))
    assert read_reply.data[0] == TEST_WORD_0

    # Demonstrate continuing to next tag, not used here.
    ex10_reader.continue_from_halted(False)
    time.sleep(0.25)  # Wait a small amount of time for reports

    while ex10_reader.packets_available():
        packet = ex10_reader.packet_peek().contents
        helper.examine_packets(packet, packet_info)
        helper.print_packets(packet)
        ex10_reader.packet_remove()
        
        if packet.packet_type == EventPacketType.TxRampDown:
            inventory_config.target = 0
            round_done = True
        elif packet.packet_type == EventPacketType.InventoryRoundSummary:
            print('Round complete')

    ex10_reader.stop_transmitting()


if __name__ == "__main__":
    run_access_rw_example()
